### DISCUSSION TITLE
Lazy load when VimEnter or edit file

### DISCUSSION THREAD

**Discussion by eeeXun** [nvim-tree/nvim-tree.lua#1954] [Q&A]
I lazy-load `nvim-tree` with following

```lua
use({
    "nvim-tree/nvim-tree.lua",
    cmd = { "NvimTreeOpen", "NvimTreeToggle" },
})
```

I want load it when enter a directory, ex. `nvim ~/`
```lua
local fn = vim.fn
local autocmd = vim.api.nvim_create_autocmd
autocmd("VimEnter", {
    callback = function()
        if fn.argc() == 1 and fn.isdirectory(fn.argv()[1]) ~= 0 then
            vim.cmd.NvimTreeOpen(fn.argv()[1])
            return true
        end
    end,
})
```
The above works for me.

I also want load it when edit a directory, ex. `:e ~/`
```lua
local fn = vim.fn
local autocmd = vim.api.nvim_create_autocmd
autocmd("BufAdd", {
    callback = function(data)
        if fn.isdirectory(data.file) ~= 0 then
            vim.cmd.NvimTreeOpen(data.file)
            return true
        end
    end,
})
```
It works when entering vim with argument (ex. `nvim some_file`), then `:e ~/`. `nvim-tree` open as expected.
But when entering vim without any argument (ex. `nvim`), so only one buffer (`[No Name]`) . Then `:e ~/`, `nvim-tree` doesn't open.

**Comment by alex-courtis** [comment]
RE: `:e ~/` have you seen `:help hijack_directories.enable` ? That will open the tree whenever you open a new directory buffer. Of course that will only function after nvim-tree setup has been called.

Honestly, I don't think lazy loading the tree is viable: there will be "race conditions" around events and their ordering.

If you are not lazy loading, you can use the API e.g. `require("nvim-tree.api").tree.toggle({ focus = false })` and rely on behaviour like `hijack_directories`.

**Comment by eeeXun** [comment]
I somehow get it works by replacing `BufAdd` with `BufEnter`.

```lua
local fn = vim.fn
local autocmd = vim.api.nvim_create_autocmd
autocmd("VimEnter", {
    callback = function()
        if fn.argc() == 1 and fn.isdirectory(fn.argv()[1]) ~= 0 then
            vim.cmd.NvimTreeOpen(fn.argv()[1])
            return true
        end
    end,
})
autocmd("BufEnter", {
    callback = function(data)
        if fn.isdirectory(data.file) ~= 0 then
            vim.cmd.NvimTreeOpen(data.file)
            return true
        end
    end,
})
```

**Reply by alex-courtis** [reply]
> I somehow get it works by replacing `BufAdd` with `BufEnter`.

That's not surprising; event ordering is practically nondeterministic.

What is you use case for lazy loading on command? nvim-tree does nothing at startup except for loading configuration and registering events - it's very cheap.

**Reply by eeeXun** [reply]
> What is you use case for lazy loading on command? nvim-tree does nothing at startup except for loading configuration and registering events - it's very cheap.

I just do it for cool. I don't have any meaningful reason. `packer` provide it, and I use it.

**Reply by alex-courtis** [reply]
Take a look at [Open At Startup](https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup) for some interesting recipes for startup.
